### DISCUSSION TITLE
Porting to ARM-based devices

### DISCUSSION THREAD

**Discussion by MangoDragonHub** [godotengine/godot-proposals#4113] [Platforms]
Hello everyone,

Out of curiosity I wanted to know if there is an official way to export your Godot game to any ARM-based device natively from the Godot editor? Like for the Raspberry Pi, and other Linux Based ARM computers. I know it supports X11 and many other x86 devices and the export settings has native Android support but when it comes to any other ARM devices I keep seeing forums that you need to have an ARM build of the Godot editor to even export to that architecture.

**Comment by Calinou** [comment]
This is already tracked in a proposal: https://github.com/godotengine/godot-proposals/issues/988

Godot can already be compiled for Linux ARM platforms, but no official editor and export template binaries are provided yet. Providing such binaries is planned for a future version, but there's no ETA as it involves nontrivial changes to the [build container](https://github.com/godotengine/build-containers) system.

**Reply by MangoDragonHub** [reply]
Can you guide me to some resources that would help me learn more about exporting to that architecture? I currently know about the un-official builds of Godot for the Raspberry Pi 4/CM4, and you can export through that hardware and using [Clickable](https://clickable.bhdouglass.com/en/latest/) to port any Godot application to Ubuntu Touch devices.

**Reply by Calinou** [reply]
The [unofficial builds](https://github.com/hiulit/Unofficial-Godot-Engine-Raspberry-Pi) are likely the  best place to start for exporting to Raspberry Pi. [frt](https://github.com/efornara/frt) can be used for some low-power devices, but it will likely give you a worse experience on a modern Raspberry Pi.

**Reply by MangoDragonHub** [reply]
Would frt be consider the optimal choice if I'm running a 3D game on a Pinebook Pro or any above average ARM laptop/singal board computer?

**Reply by Calinou** [reply]
I think it boils down to whether the Pinebook Pro has a complete (and fast) X11 setup available, or only a limited framebuffer.

**Reply by MangoDragonHub** [reply]
I'm assuming it probably does considering the chipset in that device, the RK3399s, and equivalents are used in a lot of Android SBCs. Whenever, I get far enough in development I'll see if I can get something similar enough or a get the PineBook Pro second-hand.

**Reply by ghost** [reply]
So I was just looking at this for the rpi4 because it actually has opengl es3 drivers but godot only runs on es2 on it because it uses desktop GL libraries on linux.  I think to do arm releases right that would need to change,  but I'm still a little lost looking around the build system still.

**Reply by Calinou** [reply]
Godot 3.x's GLES3 renderer is likely too demanding for a Raspberry Pi 4 (even overclocked). I'd recommend sticking to GLES2 there for performance reasons alone, since GLES2 performs better in simple scenes which you'll be running on a Raspberry Pi.

**Reply by ghost** [reply]
Interesting, thank you for that advice.  I wasn't sure about how well it would work, just knew that the driver support was there.
